### PR TITLE
[FLINK-20342][docs] Restructuring of Deployment & Operations documentation pages

### PR DESCRIPTION
**Pull request by tillrohrmann**
In order to see the new structure, please take a look at the screenshot.

Screenshot 2020-11-25 at 15 55 01

This PR also moves the configuration of the metrics reporters to `Deployment/Metrics Reporters` as it is a concern of the deployment.

cc @rmetzger

### CODE REVIEW

**Comment by XComp** [comment]
What about splitting up the Configuration page into subpages as it is quite long already. This way we could also make the Memory Configuration a subpage of Configuration

**Comment by rmetzger** [comment]
This change is already quite big. I would do such changes in follow up PRs. Ideally this change only focuses on the structure, individual pages should be updated separately.
(As a side note: I'm not convinced that splitting the Configuration page is a good idea, because then you can not search the page for config keys anymore (with CTRL+F))

File: docs/deployment/filesystems/plugins.md [hunk 1]
```diff
@@ -70,7 +70,7 @@ possible across Flink core, plugins, and user code.
 
 ## File Systems
 
-All [file systems](filesystems) **except MapR** are pluggable. That means they can and should
+All [file systems]() **except MapR** are pluggable. That means they can and should
```
Comment (rmetzger) [inline docs/deployment/filesystems/plugins.md:73]: I guess it makes sense to put a proper link here?

**Review by rmetzger — COMMENTED** [review]
I'm wondering if it makes sense to set up a redirect for the /ops/deployment/* pages, since they've all changed, leading to a lot of broken links on the internet.

**Comment by rmetzger** [comment]
Rendering of the change somehow fails locally, any ideas?

```
Configuration file: /Users/robert/Projects/flink/docs/_config.yml
            Source: /Users/robert/Projects/flink/docs
       Destination: /Users/robert/Projects/flink/docs/content
 Incremental build: disabled. Enable with --incremental
      Generating...
Is a directory @ io_fillbuf - fd:7 /Users/robert/Projects/flink/docs/deployment/memory
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/utils.rb:141:in `readline'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/utils.rb:141:in `open'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/utils.rb:141:in `has_yaml_header?'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/reader.rb:52:in `block in read_directories'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/reader.rb:48:in `each'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/reader.rb:48:in `read_directories'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/reader.rb:89:in `block in retrieve_dirs'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/reader.rb:86:in `each'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/reader.rb:86:in `retrieve_dirs'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/reader.rb:60:in `read_directories'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/reader.rb:16:in `read'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/site.rb:174:in `read'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-multiple-languages-2.0.3/lib/jekyll-multiple-languages/site.rb:19:in `read'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/site.rb:74:in `process'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-multiple-languages-2.0.3/lib/jekyll-multiple-languages/site.rb:11:in `process'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/command.rb:28:in `process_site'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/commands/build.rb:65:in `build'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/commands/build.rb:36:in `process'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/command.rb:89:in `block in process_with_graceful_fail'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/command.rb:89:in `each'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/command.rb:89:in `process_with_graceful_fail'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/lib/jekyll/commands/serve.rb:86:in `block (2 levels) in init_with_program'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `block in execute'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `each'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `execute'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/mercenary-0.3.6/lib/mercenary/program.rb:42:in `go'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/gems/jekyll-4.0.1/exe/jekyll:15:in `<top (required)>'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/bin/jekyll:23:in `load'
/Users/robert/Projects/flink/docs/.rubydeps/ruby/2.5.0/bin/jekyll:23:in `<top (required)>'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/cli/exec.rb:75:in `load'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/cli/exec.rb:75:in `kernel_load'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/cli.rb:424:in `exec'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/cli.rb:27:in `dispatch'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/cli.rb:18:in `start'
/usr/local/share/gems/gems/bundler-1.16.1/exe/bundle:30:in `block in <top (required)>'
/usr/local/share/gems/gems/bundler-1.16.1/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
/usr/local/share/gems/gems/bundler-1.16.1/exe/bundle:22:in `<top (required)>'
/usr/local/bin/bundle:23:in `load'
/usr/local/bin/bundle:23:in `<main>'                    done in 0.396 seconds.
```

**Comment by rmetzger** [comment]
From the stacktrace, it seems to be related to the multiple languages? or the has_yaml_header? maybe some header is wrong somewhere?

File: docs/deployment/filesystems/plugins.md (hunk 1 above)
Comment (tillrohrmann) [reply docs/deployment/filesystems/plugins.md:73]: Yes this makes sense.

**Comment by tillrohrmann** [comment]
Hmm, locally it builds for me. But let me check whether I can find the problem.

Concerning the redirect links, I have also thought about it. We can definitely do it but it will probably only be a best effort. Moreover, if you link to the latest documentation, then you shouldn't expect that all links stay forever the same. Maybe we should stop using `https://ci.apache.org/projects/flink/flink-docs-stable/` as the URL for the latest docs and instead use an explicit URL like `https://ci.apache.org/projects/flink/flink-docs-release-1.10/` to avoid this problem in the future.

**Comment by rmetzger** [comment]
Are you building the docs in docker, or directly in your OS? (I'm using docker)

I would only do the redirect if jekyll supports wildcards, and I would just redirect to the Deployment Overview page.

**Comment by tillrohrmann** [comment]
I am building the documents locally. One difference I've noticed is that I have Ruby 2.7.0 locally. But I guess I should use Docker for reproducibility.

**Comment by tillrohrmann** [comment]
Docker also works for me @rmetzger.

**Comment by rmetzger** [comment]
Thanks a lot for testing it with Docker as well. It was indeed a local issue. Sorry for the trouble.

**Comment by rmetzger** [comment]
Did you intentionally move the "Configuration" and "Memory Configuration" so low in the "Deployment" section. Maybe we can put it after "Resource Providers"?

**Review by rmetzger — APPROVED** [review]
Thanks a lot for this effort!

This restructuring alone is a massive improvement over the previous state of the Deployment & Operations area of the docs.
Of course the individual pages are not in a great shape yet, but this is a good foundation for further improvements.

**Comment by rmetzger** [comment]
Please remember to run the `check_links.sh` script before merging. 
Due to recent events, I don't trust my local setup anymore, but
```
≻ ./check_links.sh
http://localhost:3000/d/FOe0PbmGk/walkthrough?viewPanel=2&orgId=1&refresh=5s:
Remote file does not exist -- broken link!!!
--
http://localhost:8081/:
Remote file does not exist -- broken link!!!
--
http://localhost:4000/deployment/resource-providers/kubernetes.html:
Remote file does not exist -- broken link!!!
--
http://localhost:4000/zh/deployment/resource-providers/kubernetes.html:
Remote file does not exist -- broken link!!!
---------------------------------------------------------------------------
Found 4 broken links.
Search for page containing broken link using 'grep -R BROKEN_PATH DOCS_DIR'
```

**Comment by XComp** [comment]
Didn't we want to address YARN HA as well? It's missing in the subsection list of HA Services.

And from a stylistic point of view: IMHO, we could remove "HA Services" from "Kubernetes HA Services" and "Zookeeper HA Services" as it's obsolete considering that those are subsections of "High Availability (HA)".

**Comment by XComp** [comment]
> This change is already quite big. I would do such changes in follow up PRs. Ideally this change only focuses on the structure, individual pages should be updated separately.
> (As a side note: I'm not convinced that splitting the Configuration page is a good idea, because then you can not search the page for config keys anymore (with CTRL+F))

Searching the Configuration page is a good point. I haven't thought about that 👍

**Comment by tillrohrmann** [comment]
Yes, I agree let's remove the "HA Services" part as it is redundant.

The entries under "HA Services" are the supported HA services implementations (ZooKeeper and K8s). Hence, I think we shouldn't mention Yarn here. Instead, the Yarn resource provider page should have a subsection HA where we say what you need to configure in order to support AM restarts on Yarn which is required (next to ZooKeeper HA services) to enable master failover.

**Comment by tillrohrmann** [comment]
@XComp did you expect that `High Availability (HA) -> Kubernetes HA Services` explains how to setup HA on K8s? The idea is that this page contains explanations about the K8s HA service and how to activate it in the configuration. It won't explain which deployment yaml to change, though. If this is the case, then removing the "HA Services" might even be more misleading.

**Comment by tillrohrmann** [comment]
I actually wanted to put the Configuration page at the bottom because I thought it would make sense to close the deployment section with it (in some sense, at the end you find an overview over all configuration settings we have talked about). Given the current state where it is in the lower quarter, I guess it would make sense to move it a bit up. Will change it.

**Pull request closed without merging**